### PR TITLE
[Issue #358]: Add e2e tests against PR preview env

### PR DESCRIPTION
**Pull request by rylew1**
## Ticket

Resolves #358

## Changes

- add playwright e2e tests against preview env

## Context

- Working demo with updates from `template-infra` https://github.com/navapbc/platform-test-nextjs/pull/86
- `template-infra` PR to be merged: https://github.com/navapbc/template-infra/pull/694

## Preview environment
- Service endpoint: http://p-85-app-dev-85861930.us-east-1.elb.amazonaws.com
- Deployed commit: 1c00c002110a100cfa09454324b4abd24efbd712

### CODE REVIEW

File: e2e/playwright.config.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  /* Retry on CI only */
+  retries: process.env.CI ? 2 : 0,
+  /* Opt out of parallel tests on CI. */
+  workers: process.env.CI ? 1 : undefined,
+  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
+  reporter: "html",
+  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
+  use: {
+    /* Base URL to use in actions like `await page.goto('/')`. */
+    baseURL: process.env.CI ? undefined : "http://127.0.0.1:3000",
```
Comment (lorenyu) [inline e2e/playwright.config.js:26]: I don't like the hardcoded localhost URL here, and I also think that the fact that we set the baseURL for CI with `jq` in a script is opaque and hard to track down. I'm thinking we can be more explicit and set this via an environment variable e.g.:

```suggestion
    baseURL: process.env.BASE_URL
```

File: e2e/playwright.config.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
+    trace: "on-first-retry",
+    screenshot: "on",
+    video: "on-first-retry",
+  },
+
+  /* Configure projects for major browsers */
+  projects: [
```
Comment (lorenyu) [inline e2e/playwright.config.js:35]: question for discussion: 🤔 I'm a little skeptical that it's valuable to have the infra template's default configuration include so many browsers. What issues would this catch? I'm biased towards simplifying this to just one project configuration.

File: e2e/package.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "e2e",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test:e2e": "npx playwright test ",
+    "test:e2e:ui": "npx playwright test --ui"
```
Comment (lorenyu) [inline e2e/package.json:8]: delete? this one doesn't seem used

File: .github/workflows/e2e-tests.yml
Comment (lorenyu) [inline .github/workflows/e2e-tests.yml:1]: In order to support e2e tests for multiple apps we'll want to update this workflow to take as input an `app_name` in addition to the `service_endpoint`. Then, we'll want to generalize the steps to do something like

```
steps:
  ...
  - name: Run e2e tests
    run: npx playwright test --config ${{ inputs.app_name }}/playwright.config.js ${{ inputs.app_name }}/tests/
    env:
      BASE_URL: ${{ inputs.service_endpoint }}
```

so your directory structure will look something like

- e2e
  - playwright.config.js
  - app1/
    - playwright.config.js
    - tests/
      - index.spec.js
  - app2
    - playwright.config.js
    - tests/
      - index.spec.js

each "e2e/<APP_NAME>/playwright.config.js" can by default just import "../playwright.config.js" and re-export that file, but we can add a comment in the file saying that projects can optionally override the base playwright configuration if desired

**Review by lorenyu — COMMENTED** [review]
Amazing. So excited to have e2e tests!

File: .github/workflows/e2e-tests.yml
Comment (rylew1) [reply .github/workflows/e2e-tests.yml:1]: I think I will focus on getting this working on the a11y pr and then come back and try to fix the issues on this PR.

File: e2e/playwright.config.js (hunk 1 above)
Comment (rylew1) [reply e2e/playwright.config.js:27]: Sounds good - we set it up this way on a11y pr so let's use env in CI and locally. I think I will commit a `.env.example` and gitignore `.env` . I believe `.env` just need `BASE_URL`

File: e2e/playwright.config.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import { defineConfig, devices } from "@playwright/test";
+
+// Load environment variables from .env file if it exists
+// import * as dotenv from 'dotenv';
+// dotenv.config();
```
[the remark this answers is not in the document]
Comment (rylew1) [reply e2e/playwright.config.js]: we need dotenv for the `BASE_URL` so I uncommented these lines

File: e2e/package.json (hunk 3 above)
Comment (rylew1) [reply e2e/package.json:8]: I was using the ui all the time when developing playwright tests for the grants search - it lets you see where it's clicking on the page and such - but open to discussion if you feel strongly

image

Comment (lorenyu) [reply e2e/package.json]: oh i see, sounds good.

File: e2e/playwright.config.js (hunk 2 above)
Comment (rylew1) [reply e2e/playwright.config.js]: If this is just a base config I would propose just desktop chrome and mobile chrome as a default - and then the app specific playwright config can override with additional?

No expert here, but I noticed when developing the test spec against different browsers, Safari seemed to have some strange behavior I wasn't seeing in other browsers.

Comment (lorenyu) [reply e2e/playwright.config.js]: That works for me.

File: .github/workflows/e2e-tests.yml
Comment (rylew1) [reply .github/workflows/e2e-tests.yml:1]: Have this up working now

File: e2e/package.json [hunk 5]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "e2e",
+  "version": "1.0.0",
+  "scripts": {
+    "e2e-setup": "npx playwright install",
+    "e2e-tests": "dotenv -- cross-var npx playwright test --config %APP_NAME%/playwright.config.js",
+    "e2e-tests-ci": "cross-var npx playwright test --config %APP_NAME%/playwright.config.js",
```
Comment (rylew1) [inline e2e/package.json:7]: Not sure if there's one command for both `e2e-tests` and `e2e-tests-ci`

Comment (rylew1) [reply e2e/package.json:7]: `e2e-test` vs `e2e-tests` plural?

File: Makefile [hunk 6]
```diff
@@ -217,6 +217,21 @@ release-image-name: ## Prints the image name of the release image
...
+	@cd e2e && npm install
+	@cd e2e && npx playwright install --with-deps
+
+
+e2e-test: ## Run end-to-end tests
+    # make e2e-test APP_NAME=app BASE_URL=http://localhost:3000
+	@:$(call check_defined, APP_NAME, ...)
+	@:$(call check_defined, BASE_URL, ...)
+	@cd e2e && cd $(APP_NAME) && APP_NAME=$(APP_NAME) BASE_URL=$(BASE_URL) npx playwright test $(E2E_ARGS)
+
```
Comment (rylew1) [inline Makefile:234]: these make commands are one way to trigger the tests, but we can also trigger in the e2e folder with the e2e-tests script as well (the latter will use the env set - whether local .env or CI envs)

File: e2e/package.json (hunk 5 above)
Comment (lorenyu) [reply e2e/package.json:7]: `e2e-test` to be consistent with `npm test`, `make test`, etc

File: e2e/app/playwright.config.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { defineConfig, devices } from '@playwright/test';
+
+import baseConfig from '../playwright.config';
+
+export default defineConfig({
+  ...baseConfig,
```
Comment (lorenyu) [inline e2e/app/playwright.config.js:6]: idea:

```
use: {
  baseUrl: baseConfig.use.baseUrl || "localhost:3000"
}
```

File: .github/workflows/e2e-tests.yml [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      service_endpoint:
+        required: true
+        type: string
+      app_name:
+        required: false
+        type: string
+
+jobs:
+  e2e:
+    name: Run E2E Tests
```
Comment (lorenyu) [inline .github/workflows/e2e-tests.yml:15]:
```suggestion
    name: " " # GitHub UI is noisy when calling reusable workflows, so use whitespace for name to reduce noise
```

**Review by lorenyu — COMMENTED** [review]
in .github/workflows/ci-app-pr-environment-update.yml, rename the workflow from "CI App PR Environment Update" to "CI App PR Environment Checks" and rename the file to .github/workflows/ci-app-pr-environment-checks.yml

File: .github/workflows/e2e-tests.yml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+          libgstreamer-plugins-base1.0-0 libgstreamer-plugins-good1.0-0 libharfbuzz-icu0 libhyphen0 \
+          libenchant-2-2 libflite1 libgles2 libx264-dev
+
+      - name: Install Node.js dependencies
+        run: npm ci
+        working-directory: ./e2e
+
+      - name: Install Playwright browsers
+        run: npx playwright install --with-deps
+        working-directory: ./e2e
```
[the remark this answers is not in the document]
Comment (rylew1) [reply .github/workflows/e2e-tests.yml:44]: I would vote separate pr as this is getting large

File: .github/workflows/e2e-tests.yml [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        uses: actions/setup-node@v4
+        with:
+          node-version: '20'
+
+      - name: Install system dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y libwoff1 libopus0 libvpx7 libevent-2.1-7 libopus0 libgstreamer1.0-0 \
+          libgstreamer-plugins-base1.0-0 libgstreamer-plugins-good1.0-0 libharfbuzz-icu0 libhyphen0 \
+          libenchant-2-2 libflite1 libgles2 libx264-dev
```
[the remark this answers is not in the document]
Comment (rylew1) [reply .github/workflows/e2e-tests.yml:36]: You can see the initial failure here: https://github.com/navapbc/platform-test-nextjs/actions/runs/9885358319/job/27303354260
Playwright browsers need a boatload of deps - hope I'm not mistaken or missing something.

File: e2e/app/playwright.config.js (hunk 7 above)
Comment (rylew1) [reply e2e/app/playwright.config.js:6]: added

File: .github/workflows/e2e-tests.yml [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
```
[the remark this answers is not in the document]
Comment (rylew1) [reply .github/workflows/e2e-tests.yml:37]:
```
apt-get clean && \
    rm -rf /var/lib/apt/lists/*
```

[the remark this answers is not in the document]
Comment (lorenyu) [reply .github/workflows/e2e-tests.yml:37]: not worth it since we're not publishing the image anywhere

**Comment by rylew1** [comment]
Moved over to template-infra repo https://github.com/navapbc/template-infra/pull/694

**Comment by lorenyu** [comment]
Yep exactly. Also documented here for reference: https://github.com/navapbc/template-infra/blob/main/template-only-docs/template-development-workflow.md

**Comment by rylew1** [comment]
Closing this - see the final PR on the `template-infra` repo https://github.com/navapbc/template-infra/pull/694/files

**Pull request closed without merging**